autorest/utils/auth.go

```diff
@@ -40,3 +52,72 @@ func GetEnvVarOrExit(varName string) string {
 	}
 	return value
 }
+
+// Auth includes authentication details for ARM clients
+type Auth struct {
+	Authorizer *autorest.BearerAuthorizer
+	File       map[string]string
+	BaseURI    string
+}
+
+// GetTokenFromAuthFile creates an authorizer from an Azure CLI auth file
+func GetTokenFromAuthFile(baseURI string) (auth Auth, err error) {
+	fileLocation := os.Getenv("AZURE_AUTH_LOCATION")
+	if fileLocation == "" {
+		return auth, errors.New("auth file not found. Environment variable AZURE_AUTH_LOCATION is not set")
+	}
+
+	contents, err := ioutil.ReadFile(fileLocation)
+	if err != nil {
+		return
+	}
+
+	// Auth file might be UTF-16 LE encoded
+	decoded := decode(contents)
+
+	var af map[string]string
+	err = json.Unmarshal(decoded, &af)
+	if err != nil {
+		return
+	}
+	auth.File = af
+
+	k := getResourceKey(baseURI)
+	auth.BaseURI = af[k]
+
+	config, err := adal.NewOAuthConfig(af["activeDirectoryEndpointUrl"], af["tenantId"])
+	if err != nil {
+		return
+	}
+
+	spToken, err := adal.NewServicePrincipalToken(*config, af["clientId"], af["clientSecret"], af[k])
+	if err != nil {
+		return
+	}
+
+	auth.Authorizer = autorest.NewBearerAuthorizer(spToken)
+	return
+}
+
+func decode(b []byte) []byte {
+	utf16leBOM := []byte{255, 254}
+	if bytes.HasPrefix(b, utf16leBOM) {
+		b = bytes.TrimPrefix(b, utf16leBOM)
+
+		decoded := new(bytes.Buffer)
```

Comment:
Comment why you have to do this.

Reply:
You could also set a size since we know it will be smaller than b.

Reply:
Fixed